### ISSUE TITLE
user_functions_test.TestUserFunctions.test_restart AssertionError: Expected [['aaaaaaaaa']] from ... but got [['aaaa']]

### ISSUE THREAD

**Issue by xemul** [scylladb/scylladb#20662]
Caught while testing changes in truncation timeout passing (PR #20620)
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/599/

```
    def assert_one(session, query, expected, cl=None):
        """
        Assert query returns one row.
        @param session Session to use
        @param query Query to run
        @param expected Expected results from query
        @param cl Optional Consistency Level setting. Default ONE
    
        Examples:
        assert_one(session, "LIST USERS", ['cassandra', True])
        assert_one(session, query, [0, 0])
        """
        simple_query = SimpleStatement(query, consistency_level=cl)
        res = session.execute(simple_query)
        list_res = _rows_to_list(res)
>       assert list_res == [expected], f"Expected {[expected]} from {query}, but got {list_res}"
E       AssertionError: Expected [['aaaaaaaaa']] from SELECT sum_reduce(t) FROM tab WHERE pk in (1,3,5), but got [['aaaa']]
```

**Comment by piodul** [comment]
For the sake of clarify, this is the full stacktrace from the second failed run (https://github.com/scylladb/scylladb/issues/20662#issuecomment-2520818061):

```
self = <user_functions_test.TestUserFunctions object at 0x7f4cc2e39490>

    def test_restart(self):
        """Ensure that UDA survives server restart"""
        session = self.prepare(nodes=3)
    
        # Prepare a table with some data
        session.execute("CREATE TABLE tab (pk int PRIMARY KEY, t text)")
        for x in range(10):
            session.execute(f"INSERT INTO tab (pk, t) VALUES ({x}, '" + "a" * x + "');")
    
        # Create a UDA that sums integers from all rows
        session.execute("CREATE FUNCTION plus(acc int, val int) RETURNS NULL ON NULL INPUT RETURNS int LANGUAGE lua AS 'return acc + val'")
        session.execute("CREATE FUNCTION return_int(acc int) RETURNS NULL ON NULL INPUT RETURNS int LANGUAGE lua AS 'return acc'")
        session.execute("CREATE AGGREGATE suma(int) SFUNC plus STYPE int FINALFUNC return_int INITCOND 0")
        assert_one(session, "SELECT suma(pk) FROM tab", [45])
    
        # Create a variant of the UDA that uses a REDUCEFUNC
        session.execute("CREATE AGGREGATE sum_reduce(int) SFUNC plus STYPE int REDUCEFUNC plus FINALFUNC return_int INITCOND 0")
        assert_one(session, "SELECT sum_reduce(pk) FROM tab", [45])
    
        # Create a UDA that concatenates strings from all rows
        session.execute("CREATE FUNCTION concat(acc text, val text) RETURNS NULL ON NULL INPUT RETURNS text LANGUAGE lua AS 'return acc..val'")
        session.execute("CREATE FUNCTION return_string(acc text) RETURNS NULL ON NULL INPUT RETURNS text LANGUAGE lua AS 'return acc'")
        session.execute("CREATE AGGREGATE suma(text) SFUNC concat STYPE text FINALFUNC return_string INITCOND ''")
        assert_one(session, "SELECT suma(t) FROM tab WHERE pk IN (1,3,5)", ["aaaaaaaaa"])
    
        # Create a variant of the UDA that uses a REDUCEFUNC
        session.execute("CREATE AGGREGATE sum_reduce(text) SFUNC concat STYPE text REDUCEFUNC concat FINALFUNC return_string INITCOND ''")
>       assert_one(session, "SELECT sum_reduce(t) FROM tab WHERE pk IN (1,3,5)", ["aaaaaaaaa"])

user_functions_test.py:296: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

session = <cassandra.cluster.Session object at 0x7f4cc06acb30>
query = 'SELECT sum_reduce(t) FROM tab WHERE pk IN (1,3,5)'
expected = ['aaaaaaaaa'], cl = None

    def assert_one(session, query, expected, cl=None):
        """
        Assert query returns one row.
        @param session Session to use
        @param query Query to run
        @param expected Expected results from query
        @param cl Optional Consistency Level setting. Default ONE
    
        Examples:
        assert_one(session, "LIST USERS", ['cassandra', True])
        assert_one(session, query, [0, 0])
        """
        simple_query = SimpleStatement(query, consistency_level=cl)
        res = session.execute(simple_query)
        list_res = _rows_to_list(res)
>       assert list_res == [expected], f"Expected {[expected]} from {query}, but got {list_res}"
E       AssertionError: Expected [['aaaaaaaaa']] from SELECT sum_reduce(t) FROM tab WHERE pk IN (1,3,5), but got [['aaaa']]

tools/assertions.py:147: AssertionError
```

So it fails on a check which happens even before the node restart happens (and this test case was supposed to test that UDFs/UDAs work _after_ a restart happens).

Looks like the row for pk=5 was not accounted in the aggregation (length of the string == 4, not 9 as it was expected).

**Comment by kostja** [comment]
@temichus Please adjust the test to wait for restart to finish.

**Comment by piodul** [comment]
> [@temichus](https://github.com/temichus) Please adjust the test to wait for restart to finish.

I don't think this is a problem with the test. The test doesn't even reach the point where it starts the restart operation, it fails before that - that's what I meant to write.

My previous comment perhaps sounded like it identified the root cause of the issue, but it didn't. It's not clear to me what the issue is here, somebody would have to dig into it further, try reproducing it, and so on. But it is most likely a bug in scylla.

**Comment by Annamikhlin** [comment]
seen again on master gating
https://jenkins.scylladb.com/job/scylla-master/job/next/8716/testReport/user_functions_test/TestUserFunctions/Build___x86___dtest_with_tablets___test_restart/

```
session = <cassandra.cluster.Session object at 0x7fe02e0d0c50>
query = 'SELECT sum_reduce(t) FROM tab WHERE pk IN (1,3,5)'
expected = ['aaaaaaaaa'], cl = None

    def assert_one(session, query, expected, cl=None):
        """
        Assert query returns one row.
        @param session Session to use
        @param query Query to run
        @param expected Expected results from query
        @param cl Optional Consistency Level setting. Default ONE
    
        Examples:
        assert_one(session, "LIST USERS", ['cassandra', True])
        assert_one(session, query, [0, 0])
        """
        simple_query = SimpleStatement(query, consistency_level=cl)
        res = session.execute(simple_query)
        list_res = _rows_to_list(res)
>       assert list_res == [expected], f"Expected {[expected]} from {query}, but got {list_res}"
E       AssertionError: Expected [['aaaaaaaaa']] from SELECT sum_reduce(t) FROM tab WHERE pk IN (1,3,5), but got [['aaaaaa']]
```

**Comment by nyh** [comment]
> 
> So it fails on a check which happens even before the node restart happens (and this test case was supposed to test that UDFs/UDAs work _after_ a restart happens).
> 
> Looks like the row for pk=5 was not accounted in the aggregation (length of the string == 4, not 9 as it was expected).

Looks to me like a straightforward test bug: The test creates a 3-node cluster, and then INSERTs data into it and SELECTs it using the default consistency level which (if I understand correctly.. I'm not a dtest expert...) is **ONE**. So it is very much possible to read outdated data.

The test should write with ALL if it intends to read with ONE, or do both reads and writes with CL=QUORUM. Alternatively, the test could have used one node, not three.

**Comment by mykaul** [comment]
@piodul - based on the analysis above, let's ensure we fix the test with change of CL for read/write paths.

**Comment by nyh** [comment]
> [@piodul](https://github.com/piodul) - based on the analysis above, let's ensure we fix the test with change of CL for read/write paths.

I just came up with the same conclusion on a different test (https://github.com/scylladb/scylladb/issues/22865#issuecomment-2849057636) - to be honest I forgot about this one.

Perhaps we should find **all** the tests that have this bug. We can do this programmatically - any test that uses `patient_cql_connection` instead of `patient_exclusive_cql_connection`, and doesn't pass `consistency_level=ConsistencyLevel.LOCAL_QUORUM` to `patient_cql_connection`, is, unless I'm missing something, suspect of being flaky.

It seems these tests fail very rarely, because to fail we need to have double bad luck: the write should be slow enough to not reach one node by the time of the read, and the read should pick that one specific node that didn't yet get the write. But it seems any test that has this problem will be flaky at some point. (again, if I'm making a mistake in my analysis, feel free to correct me).

**Comment by mykaul** [comment]
The first doesn't have that many:
```
ykaul@ykaul:~/github/scylla-dtest$ git grep patient_cql_connection |grep consistency_level |grep -v QUORUM
compaction_test.py:        session = self.patient_cql_connection(node1, consistency_level=cl)
compaction_test.py:        session = self.patient_cql_connection(node1, consistency_level=cl)
cql_system_clients_test.py:        with self.patient_cql_connection(self.cluster.nodelist()[0], user="cassandra", password="cassandra", consistency_level=ConsistencyLevel.ALL) as session:
dsr/scylla_cluster/cluster.py:        with tester.patient_cql_connection(tester.cluster.nodelist()[0], request_timeout=600, consistency_level=self.validation_consistency) as session:
dsr/scylla_cluster/cluster.py:        with tester.patient_cql_connection(tester.cluster.nodelist()[0], request_timeout=600, consistency_level=ConsistencyLevel.ALL) as session:
materialized_views_test.py:        session = self.patient_cql_connection(node, consistency_level=ConsistencyLevel.ONE)
paging_test.py:        session = self.patient_cql_connection(node1, row_factory=row_factory, consistency_level=consistency_level)
secondary_indexes_test.py:        session = self.patient_cql_connection(node1, consistency_level=consistency_level, **kwargs)
```

I think ALL is also OK?

**Comment by mykaul** [comment]
test_keys_with_ttl_present() is problematic and so is test_keys_with_ttl_present from a quick look.

But of course, there are many more assert_one() that need to be looked at as well? Why is the CL there by default ONE?

**Comment by ScyllaPiotr** [comment]
> any test that uses `patient_cql_connection` or `cql_connection` (instead of `patient_exclusive_cql_connection` or `exclusive_cql_connection`), and doesn't pass `consistency_level=ConsistencyLevel.LOCAL_QUORUM` to `patient_cql_connection`, is, unless I'm missing something, suspect of being flaky.

Looking at the `dtest_setup.py::patient_cql_connection` vs `dtest_setup.py::patient_exclusive_cql_connection` I don't see why one would be susceptible to not passing `consistency_level=ConsistencyLevel.LOCAL_QUORUM` while the other one not. What am I not seeing?

**Comment by nyh** [comment]
Please correct me if I'm wrong, but I thought the "exclusive" one sends all the requests to the same coordinator, so in the simple case of n=3 rf=3, so the coordinator has the data, both reads and writes go to the same node. In the non-"exclusive" version, the different requests go to different coordinators, so you can happen to write to one coordinator and read from another, and with CL=ONE the read might not see the earlier write.

**Comment by nyh** [comment]
If I understand correctly, the non "exclusive" function uses the one node to *make* the connection, but once the connection is established the driver learns of all nodes and may use them for subsequent requests.
The "exclusive" version has a separate "white list" which specifies to only connect one of the listed nodes - and not any other node even if other nodes are present.

**Comment by m-szymon** [comment]
In case of this issue the core reason was a bug, where final merging of partial results could happen in parallel, leading to losing some results.
It happened only if aggregation function yields - and it sometimes happens in lua string concatenation (which I guess is normal).

I have no idea if I should add some kind of test for this. In dtest I triggered it by adding several thousand cql calls and running test several times. At least two nodes are required to trigger it - in unit test it would require extensive mocking and still is not that easy to always observe the issue.

There was other trace that turned out to be not relevant, but still seems suspicious:
Every string return from lua is converted to number (does it really always have to?), which results in an exception, handled, but it still triggers seastar/exception_hacks: _Unwind_RaiseException. Is it a normal behaviour or is it a fallback for situation that shouldn't happen? 

@nyh Probably not a best place anymore for further `consistency_level` discussion, but:
1. If `exclusive` version was given more nodes it also would't be 'safe', right?
2. Does connecting to the same coordinator guarantee, that consecutive write/read with CL=1 will succeed? Can't coordinator read from different replica before it wrote there? If yes, is it because it is implemented like that or it is a 'fundamental' requirement?

**Comment by nyh** [comment]
Seen again in https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8365/testReport/junit/user_functions_test/TestUserFunctions/FullDtest___full_split001___test_restart/

**Comment by nyh** [comment]
> In case of this issue the core reason was a bug, where final merging of partial results could happen in parallel, leading to losing some results. It happened only if aggregation function yields - and it sometimes happens in lua string concatenation (which I guess is normal).

Good catch! We need to get your fix https://github.com/scylladb/scylladb/issues/20662 reviewed quickly, because this is breaking CI too much (just saw it again 3 minutes ago) :-(
 
> I have no idea if I should add some kind of test for this. In dtest I triggered it by adding several thousand cql calls and running test several times. At least two nodes are required to trigger it - in unit test it would require extensive mocking and still is not that easy to always observe the issue.

It might be possible to write a test for it in test/cluster. If the test has a loop of 1000 calls and takes a few seconds, and has a, say, 50% chance of reproducing the bug, it's still better than nothing, because you can see
But I don't know if it's worth the effort.
We can also write a test with "injections" - to add some sort of artificial slowness or yield or something to a specific place in the aggregation function, that's also a possiblity (but I don't know if it's one I recommend).

> There was other trace that turned out to be not relevant, but still seems suspicious: Every string return from lua is converted to number (does it really always have to?), which results in an exception, handled, but it still triggers seastar/exception_hacks: _Unwind_RaiseException. Is it a normal behaviour or is it a fallback for situation that shouldn't happen?

I don't know, doesn't sound normal. This sounds like something you should be able to reproduce in the single-node test/cqlpy, and if you can reproduce a bug in that you can open a new issue about it.

> [@nyh](https://github.com/nyh) Probably not a best place anymore for further `consistency_level` discussion, but:
> 
>     1. If `exclusive` version was given more nodes it also wouldn't be 'safe', right?
> 
>     2. Does connecting to the same coordinator guarantee, that consecutive write/read with CL=1 will succeed? Can't coordinator read from different replica before it wrote there? If yes, is it because it is implemented like that or it is a 'fundamental' requirement?

We've been discussing this consistency-level issue across multiple issues so I'm a bit lost as to what is relevant in this issue and what isn't. Basically, in a test with 3 nodes and RF=3 (many of our dtests use this, as some sort of "standard cluster"), each one of the nodes has a copy of the data, so if you contact a specific coordinator, it will always read its own data. So if you have an "exclusive" connection you'll always read this data from this one coordinator, and also writes will return after writing it, but if you have a non-"exclusive" connection, you can read or write to different coordinators, and possibly not read your own writes.
This is all for reading with CL=1. If you read with QUORUM, and write with QUORUM, you're always able to read what you wrote.

If I remember correctly, when I reviewed this specific test, I believed it has this problem (CL=1, non-exclusive connection). Of course it doesn't mean there's any chance to actually see this problem in practice. To see this problem we need the write to finish on just one node and not the two other nodes - and the following read will reach exactly those two nodes and read from them before the write finished. The chance of this happening are probably very slim, so we never saw this problem in practice. I guess the problem that you find is more likely to happen.
